console/v1/types_console_plugin.go

```diff
@@ -48,6 +48,131 @@ type ConsolePluginSpec struct {
+)
+
+// CSPDirectiveValue is single value for a Content-Security-Policy directive.
+// Each directive value must have a maximum length of 1024 characters and must not contain
+// whitespace, commas, or semicolons. The value '*' is not permitted.
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=1024
+// +kubebuilder:validation:XValidation:rule="!self.contains(\"'\")",message="CSP directive value cannot contain a quote"
+// +kubebuilder:validation:XValidation:rule="!self.matches('\\\\s')",message="CSP directive value cannot contain a whitespace"
+// +kubebuilder:validation:XValidation:rule="!self.contains(',')",message="CSP directive value cannot contain a comma"
+// +kubebuilder:validation:XValidation:rule="!self.contains(';')",message="CSP directive value cannot contain a semi-colon"
+// +kubebuilder:validation:XValidation:rule="self != '*'",message="CSP directive value cannot be a wildcard"
+type CSPDirectiveValue string
+
+// ConsolePluginCSP holds configuration for a specific CSP directive
+type ConsolePluginCSP struct {
+	// directive specifies which Content-Security-Policy directive to configure.
+	// Available directive types are DefaultSrc, ScriptSrc, StyleSrc, ImgSrc and FontSrc.
+	// DefaultSrc directive serves as a fallback for the other CSP fetch directives.
+	// For more information about the DefaultSrc directive, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/default-src
+	// ScriptSrc directive specifies valid sources for JavaScript.
+	// For more information about the ScriptSrc directive, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src
+	// StyleSrc directive specifies valid sources for stylesheets.
+	// For more information about the StyleSrc directive, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src
+	// ImgSrc directive specifies a valid sources of images and favicons.
+	// For more information about the ImgSrc directive, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/img-src
+	// FontSrc directive specifies valid sources for fonts loaded using @font-face.
+	// For more information about the FontSrc directive, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/font-src
+	// +kubebuilder:validation:Required
+	Directive DirectiveType `json:"directive"`
+	// values defines an array of values to append to the console defaults for this directive.
+	// Each ConsolePlugin may define their own directives with their values. These will be set
+	// by the OpenShift web console's backend, as part of its Content-Security-Policy header.
+	// The array can contain at most 16 values. Each directive value must have a maximum length
+	// of 1024 characters and must not contain whitespace, commas, semicolons. The value '*' or
+	// strings enclosed in single quotes are not permitted.
```

Comment:
We've changed this validation slightly since this was written
```suggestion
	// The array can contain at most 16 values. Each directive value must have a maximum length
	// of 1024 characters and must not contain whitespace, commas (,), semicolons (;) or single quotes (').
	// The value '*' is not permitted.
```

Reply:
@JoelSpeed API looks good to me once we address this last comment. Thanks, Joel.